src/node.rs

```diff
@@ -709,32 +774,41 @@ impl KademliaNode {
 
     pub async fn ping(&self, addr: SocketAddr) -> Result<(), KademliaError> {
         debug!("Pinging node at {} from {}", addr, self.addr);
-        self.network_manager
-            .send_message(&Message::Ping { sender: self.id }, addr)
-            .await?;
 
-        match tokio::time::timeout(
-            self.config.request_timeout,
-            self.network_manager.receive_message(),
+        exponential_backoff(
+            || {
+                Box::pin(async move {
+                    self.network_manager
+                        .send_message(&Message::Ping { sender: self.id }, addr)
+                        .await?;
+
+                    match tokio::time::timeout(
+                        self.config.request_timeout,
+                        self.network_manager.receive_message(),
+                    )
+                    .await
+                    {
+                        Ok(Ok((Message::Pong { sender }, src))) => {
```

Comment:
nested matching like so is not as legible as just doing separate `match` statements.